ControllerDriver/networkhandler.cpp

```diff
@@ -0,0 +1,108 @@
+#include "networkhandler.h"
+#include "queue.h"
+#include <WiFi.h>
+#include "networkconfig.h"
+#include "main.h"
+
+const char* ssid     = NETSSID;
+const char* password = NETPASS;
+const char* host = NETHOST;
+
+const int serverPort = 11000;
+
+unsigned long waitUntil = 0;
+
+// Use WiFiClient class to create TCP connections
+WiFiClient client;
+
+int wStatus = WL_IDLE_STATUS;
+QueueArray<struct Packet*> buffer;
+
+void handleNetworkStuff() {
+  if(waitUntil < getLoopTime()) {
+    wStatus = WiFi.status();
+  
+    if(wStatus != WL_CONNECTED && wStatus != WL_IDLE_STATUS) {
+      WiFi.begin(ssid, password);
+      Serial.print("Current status: ");
+      Serial.print(wStatus);
+      Serial.print(". Reconnecting to ");
+      Serial.println(ssid);
+      waitUntil = getLoopTime() + 500;
+    }
+    
+    if(wStatus == WL_CONNECTED) {
+      if(client.connected()) {
+        Packet* p = buffer.pop();
+        if(p != NULL) {
+          int sent = client.write(p->data, p->length);
+          Serial.print("Bytes sent: ");
+          Serial.println(sent);
+          delete[] p->data;
+          delete p;
+        }
+      
+        // Read all the lines of the reply from server and print them to Serial
+        while (client.available()) {
+          String line = client.readStringUntil('\0');
+          Serial.print(line);
+        }
+      } else {
+        if(client.connect(host, serverPort)) {
+          Serial.print("Connected to ");
+          Serial.println(host);
+        } else {
+          Serial.print("Failed to connect to ");
+          Serial.println(host);
+        }
+      }
+    }
+  }
+}
+
+void pushNetworkPacket(Packet* packet) {
+  buffer.enqueue(packet);
+}
+
+void makeNetworkPacket(const uint8_t &type, const uint8_t* contents, uint16_t contentLength, const uint8_t &priority) {
+  Packet* p = new Packet();
+  
+  uint16_t pSize = contentLength + 5;
+  p->data = new uint8_t[pSize]; // 2 for ||, 2 for length, 1 for type.
+  p->data[0] = '|';
+  writeNumberToBuffer<uint16_t>(contentLength, p->data, 1);
+  writeNumberToBuffer<uint8_t>(type, p->data, 3);
+  memcpy(&p->data[4], contents, contentLength);
+  p->data[pSize - 1] = '|';
+  p->length = pSize;
+  p->priority = priority;
+  pushNetworkPacket(p);
+}
+
+bool isConnected() {
+  return (wStatus == WL_CONNECTED);
+}
+
+template <typename T>
+void swapEndian(T &val) {
```

Comment:
Also, changing endianness for network packets is a common problem.  A common solution is to use `htonl` and `htons` to convert values where needed ([example on StackOverflow](https://stackoverflow.com/questions/5795283/creating-and-sending-data-packets-in-c-c)).  You should be able to replace your `swapEndianIfNeeded` calls pretty easily.

Reply:
I don't like htons and htonl because, much like the other network stuff, they vary by platform, and they're not generic. They only support uint16_t and uint32_t. The platform argument doesn't matter for the ESP32-Cam, but we also need to do endian changing on the server so we still need a good function for this.

I found this example code online. Finding a better way to do it is probably a matter for a following PR, or a discussion on Discord.

Reply:
Resolving this conversation because I changed it to htons/htonl.